api/src/main/java/org/openmrs/module/appointmentscheduling/api/impl/AppointmentServiceImpl.java

```diff
@@ -1155,6 +1167,117 @@ public Appointment bookAppointment(Appointment appointment,
 				appointment);
 	}
 
+
+	//appointment resource
+
+	@Override
+	public List<AppointmentResource> getAllAppointmentResources() {
+		return getAppointmentResourceDAO().getAll();
+	}
+
+	@Override
+	public List<AppointmentResource> getAllAppointmentResources(boolean includeVoided) {
+		return getAppointmentResourceDAO().getAllData(includeVoided);
+	}
+
+	@Override
+	public AppointmentResource getAppointmentResource(Integer appointmentResourceId) {
+		return (AppointmentResource) getAppointmentResourceDAO().getById(appointmentResourceId);
+	}
+
+	@Override
+	public AppointmentResource getAppointmentResourceByUuid(String uuid) {
+		return (AppointmentResource) getAppointmentResourceDAO().getByUuid(uuid);
+	}
+
+	@Override
+	public AppointmentResource saveAppointmentResource(AppointmentResource appointmentResource) throws APIException {
+		ValidateUtil.validate(appointmentResource);
+		return (AppointmentResource) getAppointmentResourceDAO().saveOrUpdate(appointmentResource);
+	}
+
+	@Override
+	public AppointmentResource voidAppointmentResource(AppointmentResource appointmentResource, String reason) {
+		appointmentResource.setVoided(true);
+		appointmentResource.setVoidReason(reason);
+		return (AppointmentResource) getAppointmentResourceDAO().saveOrUpdate(appointmentResource);
+	}
+
+	@Override
+	public void purgeAppointmentResource(AppointmentResource appointmentResource) {
+		getAppointmentResourceDAO().delete(appointmentResource);
+	}
+
+	@Override
+	public List<AppointmentResource> getAppointmentResourcesByConstraints(Location location, Provider provider, List<AppointmentType> appointmentTypes) {
+		return getAppointmentResourceDAO().getResourceByConstraints(location, provider,null);
+	}
+
+	@Override
+	public TimeSlot getTimeslotForAppointment(Location location, Provider provider, AppointmentType type, Date appointmentDate){
+		TimeSlot requiredTimeslot = null;
+
+		Date dayBeforeAppointmentDate = getDate(appointmentDate, -1);
+		Date dayAfterAppointmentDate = getDate(appointmentDate, 1);
+
+		List<TimeSlot> timeSlots = getTimeSlotsByConstraintsIncludingFull(type, dayBeforeAppointmentDate, dayAfterAppointmentDate, provider, location);
+
+		for (TimeSlot timeSlot : timeSlots){
+			if ((timeSlot.getStartDate().getTime() / 1000 <= appointmentDate.getTime() / 1000) && ( appointmentDate.getTime() / 1000 <= timeSlot.getEndDate().getTime() / 1000)){
```

Comment:
So I think this is the logic where if htere are multiple timeslots for a provider on the current day, it decides which one to use?  Can you explain what this logic does and why it works this way? Sorry, it's a little confusing to me!

Reply:
Loops through the list of timeslots and select the one where
timeslot start time <= apointment time and timeslot endtime >= appointment time

appointment time is the time passed by the user

Reply:
I converted the times to unix timestamp for easy comparison

Reply:
I would think the getTimeSlotsByConstraints should *only* return timeslots where timeslot start time <= apointment time and timeslot endtime <= appointment time...

So it looks like this method would just loop through all the returned timeslots and requiredTimeslot would just end up being the last one set... couldn't you just do something like:

```
List<TimeSlot> timeSlots = getTimeSlotsByConstraintsIncludingFull(type, setDateToStartOfDay(appointmentDate), setDateToEndOfDay(appointmentDate), provider, location);

if (timeSlots.size > 0) {
  return timeSlots.get(0)
}
else {
  return null;
}
```

.. and then be done with it?

Reply:
that's not what's happening in that loop

Reply:
i'm using this getTimeSlotsByConstraints to return all timeslots of the day
lets say they are two 7:00:00 - 12:00:00, 13:00:00 - 18:00:00.
The loop will return the appropriate timeslot to the appointment time after iterating the timeslot list with the above two timeslot, let's say appointment time is 10:00:00
it will return the first timeslot because 7:00:00 <= 10:00:00 <= 12:00:00 evaluates to a boolean TRUE. If appointment time is 13:00:00 it will return the second timeslot

so this "So it looks like this method would just loop through all the returned timeslots and requiredTimeslot would just end up being the last one set" is not what's happening

Reply:
Thakns @jecihjoy  I see I bug in my sample code avoid... I see how if f you use getTimeSlotsByConstraints to fetch all the time slots on a current day, you'd need the logic you added to pick the correct one...

I think my mistake was that I was hoping that getTimeSlowsByConstraints would be smart enough to handle the "time" bit for you this for you... that you could do something like:

`List<TimeSlot> timeSlots = getTimeSlotsByConstraintsIncludingFull(type, appointmentDate, appointmentDate, provider, location)`

... and in the example you give above, if the appoinmentDate had a time component of "10:00" the getTimeSlotsByConstraints would just return the '7:00:00 - 12:00:00" time slot... (Note that it the above I just pass in the appointment date without setting it to start of day and end of day... ) 

If getTimeSlotsByConstraints doesn't work this way, then, yes, I understand you'd need additional logic that you added so you get the right timeslot... I just wanted to make sure we confirm that getTimeSlotsByConstraints won't handle time components properly... and couldn't be easily modified to do so.It seems like this logic belongs in the getTimeSlotByConstraints method, but if there's a reason it wouldn't work or it's a lot of work to fix I'm fine the way you are doing it... hope that makes sense!

Reply:
Ah, okay, I see your comment from above, which sounds like getTimeSlotsByConstraints *can't* do what I was hoping for above:

"This is not likely to work because the dates carry both the date and time and
1 . getTimeSlotByConstraints compares using both the date and time,
fromDate==toDate will never return true because although the dates might be the same the start time and end time of an appointment are never equal
2. Formatting the dates to exclude time in getTimeSlotByConstraints will return more than one timeslot because a day can have multiple timeslot"

If there's not easy tweak to getTimeSlotByConstraints to make it do what you want, then, yeah, let's go with what you have.

Reply:
I support we don't tweak the  getTimeSlotByConstraints in this case because we don't want to affect the functionalities that are using it